query-planner-js/src/composedSchema/buildComposedSchema.ts

```diff
@@ -0,0 +1,240 @@
+      type.astNode,
+    );
+
+    const typeMetadata: FederationTypeMetadata = ownerDirectiveArgs
+      ? {
+          serviceName: graphMap[ownerDirectiveArgs?.['graph']].serviceName,
+          keys: new MultiMap(),
+          isValueType: false,
+        }
+      : {
+          isValueType: true,
+        };
+
+    type.extensions = {
+      ...type.extensions,
+      federation: typeMetadata,
+    };
+
+    const typeDirectivesArgs = getArgumentValuesForRepeatableDirective(
+      typeDirective,
+      type.astNode,
+    );
+
+    assert(
+      !(typeMetadata.isValueType && typeDirectivesArgs.length >= 1),
+      `GraphQL type "${type.name}" cannot have a @${typeDirective.name} \
+directive without an $${ownerDirective.name} directive`,
+    );
+
+    for (const typeDirectiveArgs of typeDirectivesArgs) {
+      const serviceName = graphMap[typeDirectiveArgs['graph']].serviceName;
+
+      const keyFields = parseFieldSet(typeDirectiveArgs['key']);
+
+      typeMetadata.keys?.add(serviceName, keyFields);
+    }
+
+    for (const fieldDef of Object.values(type.getFields())) {
+      assert(
+        fieldDef.astNode,
+        `Field "${type.name}.${fieldDef.name}" should contain AST nodes`,
+      );
+
+      const fieldDirectiveArgs = getArgumentValuesForDirective(
+        fieldDirective,
+        fieldDef.astNode,
+      );
+
+      if (!fieldDirectiveArgs) continue;
+
+      const fieldMetadata: FederationFieldMetadata = {
+        serviceName: graphMap[fieldDirectiveArgs?.['graph']]?.serviceName,
+      };
+
+      fieldDef.extensions = {
+        ...fieldDef.extensions,
+        federation: fieldMetadata,
+      };
+
+      if (fieldDirectiveArgs) {
+        const { requires, provides } = fieldDirectiveArgs;
+
+        if (requires) {
+          fieldMetadata.requires = parseFieldSet(requires);
+        }
+
+        if (provides) {
+          fieldMetadata.provides = parseFieldSet(provides);
+        }
+      }
+    }
+  }
+
+  // We filter out schema elements that should not be exported to get to the
+  // API schema.
+
+  const schemaConfig = schema.toConfig();
+
+  return new GraphQLSchema({
+    ...schemaConfig,
+    types: schemaConfig.types.filter(isExported),
+    directives: schemaConfig.directives.filter(isExported),
+  });
+
+  // TODO: Implement the IsExported algorithm from the Core Schema spec.
+  function isExported(element: NamedSchemaElement) {
+    return !(isAssociatedWithFeature(element, coreName) || isAssociatedWithFeature(element, joinName))
+  }
+
+  function isAssociatedWithFeature(
+    element: NamedSchemaElement,
+    featureName: string,
+  ) {
+    return (
+      element.name === `${featureName}` ||
+      element.name.startsWith(`${featureName}__`)
+    );
+  }
+}
+
+type NamedSchemaElement = GraphQLDirective | GraphQLNamedType;
+
+function parseFieldSet(source: string): FieldSet {
+  const selections = parseSelections(source);
+
+  assert(
+    selections.every(isASTKind('Field', 'InlineFragment')),
```

Comment:
Should this be (at least in theory, perhaps a follow-up PR) a recursive check rather than just a top-level one? (using the visitor or something)

Reply:
Yes, that would be a better approach. I'll save that for a follow-up PR.